src/cpp/session/modules/SessionLinter.R

```diff
@@ -13,146 +13,124 @@
+.rs.addFunction("lint", function(filePath)
 {
    engine <- .rs.getLintEngine()
-   engine(filePath)
+   invisible(engine(filePath))
 })
 
-.rs.addFunction("internalLinter", function(filePath)
+.rs.addFunction("availableRSymbols", function()
+{
+   unlist(c(
+      .rs.objectsOnSearchPath(excludeGlobalEnv = TRUE),
+      .rs.getVar("r.keywords")
+   ))
+})
+
+.rs.addFunction("internalLintEngine", function(filePath)
 {
-   
    if (!file.exists(filePath))
       return(list())
-   
-   parsed <- tryCatch(
-      parse(file = filePath),
-      error = function(e) return(e)
+      
+   filePath <- .rs.normalizePath(filePath, mustWork = TRUE)
+   lint <- .rs.lintRFile(filePath)
+   invisible(.rs.showLintMarkers(lint, filePath))
+})
+
+.rs.addFunction("lintRFile", function(filePath)
+{
+   .Call("rs_lintRFile", filePath)
+})
+
+.rs.addFunction("showLintMarkers", function(lint, filePath)
+{
+   markers <- .rs.createMarkersFromLint(lint, filePath)
+   .rs.api.sourceMarkers(
+      name = "Linter",
+      markers = markers,
+      basePath = .rs.getProjectDirectory()
    )
-   
-   if (inherits(parsed, "simpleError"))
-   {
-      splat <- strsplit(parsed$message, "\n", perl = TRUE)[[1]]
-      row <- as.integer(gsub(":.*", "", splat[[2]]))
-      column <- c(regexpr("^", splat[[3]], fixed = TRUE))
-      text <- gsub(".*:\\s*", "", splat[[1]])
-      return(list(.rs.makeLintObject(
-         rowStart = row,
-         columnStart = column,
-         text = text,
-         type = "error"
-      )))
-   }
-   
-   # Construct a scope tree from the parse tree -- ie, closure scopes
-   parsed <- parse(text = "a <- 1; c(function(d) a + .__nothing__. + d + 1, b, c)")
-   root <- new.env(parent = emptyenv())
-   availableSymbols <- unlist(.rs.objectsOnSearchPath())
-   .rs.populateLintTree(parsed, root, availableSymbols)
-   
 })
 
-.rs.addFunction("getVariablesInScope", function(node) {
-   
-   vars <- character()
-   empty <- emptyenv()
-   
-   parent <- node
-   while (!identical(parent, empty))
-   {
-      vars <- c(vars, parent$local.variables)
-      if (length(parent$arguments))
-         vars <- c(vars, names(parent$arguments))
-      parent <- parent.env(parent)
-   }
-   
-   vars
+.rs.addFunction("createMarkersFromLint", function(lint, file) {
+   lapply(lint, function(x) {
+      list(
+         type = x$type,
+         file = file,
+         line = x$start.row,
+         column = x$start.column,
+         message = x$message,
+         messageHTML = FALSE
+      )
+   })
 })
 
-.rs.addFunction("populateLintTree", function(object, node, availableSymbols)
+.rs.addFunction("extractRCode", function(content,
+                                         reStart,
+                                         reEnd)
 {
-   if (is.expression(object))
-      lapply(object, function(x) .rs.populateLintTree(x, node, availableSymbols))
+   splat <- strsplit(content, "\n", fixed = TRUE)[[1]]
+   starts <- grep(reStart, splat, perl = TRUE)
```

Comment:
This reminds me that we should do a check to make sure all of our R code still works in R 2.11.1